File: services/umbra/src/routing/module-router.ts
import axios from 'axios';
import { Envelope, UmbraPayload, ModuleResult, SERVICES, API_ENDPOINTS } from '@umbra/shared';
import { RetryUtils } from '@umbra/shared';
import Logger from '@umbra/shared/src/utils/logger';

export class ModuleRouter {
  private logger: Logger;
  private retryConfig: any;

  constructor() {
    this.logger = new Logger('ModuleRouter');
    this.retryConfig = RetryUtils.createRetryConfig('api');
  }

  async routeToModule(envelope: Envelope<UmbraPayload>): Promise<ModuleResult<any>> {
    const { reqId, payload } = envelope;
    const targetModule = payload.targetModule;

    if (!targetModule || targetModule === 'umbra') {
      return {
        reqId,
        status: 'error',
        error: {
          type: 'functional',
          code: 'NO_TARGET_MODULE',
          message: 'No target module specified or routing to self',
          retryable: false
        }
      };
    }

    const startTime = Date.now();

    try {
      this.logger.debug('Routing envelope to module', {
        reqId,
        targetModule,
        action: payload.action
      });

      const result = await this.sendToModule(targetModule, envelope);
      const durationMs = Date.now() - startTime;

      this.logger.audit('Module routing completed', envelope.userId, {
        reqId,
        targetModule,
        status: result.status,
        durationMs
      });

      return result;

    } catch (error) {
      const durationMs = Date.now() - startTime;
      
      this.logger.error('Module routing failed', {
        reqId,
        targetModule,
        error: error.message
      });

      return {
        reqId,
        status: 'error',
        error: {
          type: 'technical',
          code: 'ROUTING_ERROR',
          message: `Failed to route to module: ${targetModule}`,
          retryable: true
        },
        audit: {
          module: 'umbra-router',
          durationMs
        }
      };
    }
  }

  private async sendToModule(moduleName: string, envelope: Envelope<any>): Promise<ModuleResult<any>> {
    const service = SERVICES[moduleName];
    
    if (!service) {
      throw new Error(`Unknown module: ${moduleName}`);
    }

    // Build the target URL
    const baseUrl = this.getModuleBaseUrl(moduleName);
    const endpoint = this.getModuleEndpoint(moduleName, envelope.payload.action);
    const url = `${baseUrl}${endpoint}`;

    return RetryUtils.retry(async () => {
      const response = await axios.post(url, envelope, {
        headers: {
          'Content-Type': 'application/json',
          'X-API-Key': this.getModuleApiKey(moduleName),
          'X-Service-Name': 'umbra'
        },
        timeout: 30000
      });

      return response.data;
    }, this.retryConfig);
  }

  private getModuleBaseUrl(moduleName: string): string {
    // In development, use localhost
    if (process.env.NODE_ENV === 'development') {
      const port = SERVICES[moduleName].port;
      return `http://localhost:${port}`;
    }

    // In production, use Railway internal URLs or configured URLs
    const envKey = `${moduleName.toUpperCase()}_URL`;
    const configuredUrl = process.env[envKey];
    
    if (configuredUrl) {
      return configuredUrl;
    }

    // Fallback to Railway internal service naming
    return `http://${moduleName}:${SERVICES[moduleName].port}`;
  }

  private getModuleEndpoint(moduleName: string, action: string): string {
    const endpoints = API_ENDPOINTS[moduleName as keyof typeof API_ENDPOINTS];
    
    if (!endpoints) {
      return '/api/v1/process'; // Default endpoint
    }

    // Map actions to specific endpoints
    switch (moduleName) {
      case 'finance':
        if (action === 'ocr' || action === 'extract') return endpoints.ocr;
        if (action === 'report') return endpoints.report;
        return endpoints.extract;
        
      case 'concierge':
        if (action === 'monitor') return endpoints.monitor;
        if (action === 'validate') return endpoints.validate;
        return endpoints.execute;
        
      case 'business':
        if (action === 'manage_client') return endpoints.clients;
        return endpoints.delegate;
        
      case 'production':
        if (action === 'deploy') return endpoints.deploy;
        return endpoints.workflow;
        
      case 'creator':
        if (action.startsWith('generate')) return endpoints.generate;
        return endpoints.media;
        
      case 'mcp':
        if (action === 'validate') return endpoints.validate;
        return endpoints.lifecycle;
        
      default:
        return '/api/v1/process';
    }
  }

  private getModuleApiKey(moduleName: string): string {
    const envKey = `${moduleName.toUpperCase()}_API_KEY`;
    const apiKey = process.env[envKey];
    
    if (!apiKey) {
      this.logger.warn('Missing API key for module', { moduleName, envKey });
      throw new Error(`Missing API key for module: ${moduleName}`);
    }
    
    return apiKey;
  }

  /**
   * Check if a module is available
   */
  async checkModuleHealth(moduleName: string): Promise<boolean> {
    try {
      const baseUrl = this.getModuleBaseUrl(moduleName);
      const healthEndpoint = SERVICES[moduleName].health;
      
      const response = await axios.get(`${baseUrl}${healthEndpoint}`, {
        timeout: 5000,
        headers: {
          'X-Service-Name': 'umbra'
        }
      });

      return response.status === 200;
    } catch (error) {
      this.logger.warn('Module health check failed', {
        moduleName,
        error: error.message
      });
      return false;
    }
  }

  /**
   * Get available modules and their status
   */
  async getModuleStatus(): Promise<Record<string, { available: boolean; latency?: number }>> {
    const status: Record<string, { available: boolean; latency?: number }> = {};

    const checks = Object.keys(SERVICES).map(async (moduleName) => {
      if (moduleName === 'umbra') return; // Skip self

      const startTime = Date.now();
      const available = await this.checkModuleHealth(moduleName);
      const latency = available ? Date.now() - startTime : undefined;

      status[moduleName] = { available, latency };
    });

    await Promise.all(checks);
    return status;
  }

  /**
   * Route with fallback logic
   */
  async routeWithFallback(envelope: Envelope<UmbraPayload>): Promise<ModuleResult<any>> {
    const { reqId, payload } = envelope;
    const primaryModule = payload.targetModule;

    if (!primaryModule || primaryModule === 'umbra') {
      return this.routeToModule(envelope);
    }

    // Check if primary module is available
    const isAvailable = await this.checkModuleHealth(primaryModule);
    
    if (!isAvailable) {
      this.logger.warn('Primary module unavailable, checking fallbacks', {
        reqId,
        primaryModule
      });

      // Get fallback module based on intent
      const fallbackModule = this.getFallbackModule(payload.intent || '');
      
      if (fallbackModule && fallbackModule !== primaryModule) {
        const fallbackAvailable = await this.checkModuleHealth(fallbackModule);
        
        if (fallbackAvailable) {
          this.logger.info('Using fallback module', {
            reqId,
            primaryModule,
            fallbackModule
          });

          // Update envelope to use fallback
          const fallbackEnvelope = {
            ...envelope,
            payload: {
              ...payload,
              targetModule: fallbackModule
            }
          };

          return this.routeToModule(fallbackEnvelope);
        }
      }

      // No fallback available
      return {
        reqId,
        status: 'error',
        error: {
          type: 'technical',
          code: 'MODULE_UNAVAILABLE',
          message: `Module ${primaryModule} is currently unavailable`,
          retryable: true
        }
      };
    }

    // Primary module is available, route normally
    return this.routeToModule(envelope);
  }

  private getFallbackModule(intent: string): string | null {
    // Simple fallback logic - could be made more sophisticated
    const fallbacks: Record<string, string> = {
      'finance_ocr': 'umbra', // Fall back to Umbra for basic text extraction
      'translation': 'umbra', // Umbra can handle simple translation
      'calculation': 'umbra', // Umbra can handle basic math
    };

    return fallbacks[intent] || null;
  }
}

File: python_services/umbra/umbra_service/routing/module_router.py
"""Module router for forwarding requests to appropriate services."""

import httpx
import os
from typing import Dict, Any, Optional
from datetime import datetime

from umbra_shared import (
    UmbraLogger,
    RetryUtils,
    RetryConfig,
    Envelope,
    UmbraPayload,
    ModuleResult,
    retry_async,
)


class ModuleRouter:
    """Route requests to appropriate Umbra modules."""
    
    def __init__(self):
        self.logger = UmbraLogger("ModuleRouter")
        self.retry_utils = RetryUtils()
        self.retry_config = RetryUtils.create_retry_config("api")
        
        # Module configuration
        self.modules = {
            "finance": {
                "base_url": os.getenv("FINANCE_BASE_URL", "http://localhost:8081"),
                "api_key": os.getenv("FINANCE_API_KEY"),
                "health_endpoint": "/health",
                "api_endpoint": "/api/v1"
            },
            "concierge": {
                "base_url": os.getenv("CONCIERGE_BASE_URL", "http://localhost:9090"),
                "api_key": os.getenv("CONCIERGE_API_KEY"),
                "health_endpoint": "/health",
                "api_endpoint": "/api/v1"
            },
            "business": {
                "base_url": os.getenv("BUSINESS_BASE_URL", "http://localhost:8082"),
                "api_key": os.getenv("BUSINESS_API_KEY"),
                "health_endpoint": "/health",
                "api_endpoint": "/api/v1"
            },
            "production": {
                "base_url": os.getenv("PRODUCTION_BASE_URL", "http://localhost:8083"),
                "api_key": os.getenv("PRODUCTION_API_KEY"),
                "health_endpoint": "/health",
                "api_endpoint": "/api/v1"
            },
            "creator": {
                "base_url": os.getenv("CREATOR_BASE_URL", "http://localhost:8084"),
                "api_key": os.getenv("CREATOR_API_KEY"),
                "health_endpoint": "/health",
                "api_endpoint": "/api/v1"
            }
        }
    
    async def route_to_module(
        self,
        envelope: Envelope[UmbraPayload]
    ) -> ModuleResult[Any]:
        """Route envelope to the appropriate module."""
        start_time = datetime.utcnow()
        req_id = envelope.req_id
        target_module = envelope.payload.target_module
        
        if not target_module:
            return ModuleResult(
                req_id=req_id,
                status="error",
                error={
                    "type": "functional",
                    "code": "NO_TARGET_MODULE",
                    "message": "No target module specified in payload",
                    "retryable": False
                }
            )
        
        if target_module not in self.modules:
            return ModuleResult(
                req_id=req_id,
                status="error",
                error={
                    "type": "functional",
                    "code": "INVALID_MODULE",
                    "message": f"Invalid target module: {target_module}",
                    "retryable": False
                }
            )
        
        try:
            self.logger.info("Routing to module",
                           req_id=req_id,
                           target_module=target_module)
            
            # Check module health first
            is_healthy = await self.check_module_health(target_module)
            if not is_healthy:
                self.logger.warning("Target module unhealthy, trying anyway",
                                  req_id=req_id,
                                  target_module=target_module)
            
            # Send to module
            result = await self._send_to_module(target_module, envelope)
            
            duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
            
            self.logger.info("Module routing successful",
                           req_id=req_id,
                           target_module=target_module,
                           status=result.get("status"))
            
            return ModuleResult(
                req_id=req_id,
                status=result.get("status", "success"),
                data=result.get("data"),
                error=result.get("error"),
                audit={
                    "module": "umbra-router",
                    "duration_ms": duration_ms,
                    "target_module": target_module
                }
            )
            
        except Exception as e:
            duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
            
            self.logger.error("Module routing failed",
                            req_id=req_id,
                            target_module=target_module,
                            error=str(e))
            
            return ModuleResult(
                req_id=req_id,
                status="error",
                error={
                    "type": "technical",
                    "code": "ROUTING_ERROR",
                    "message": f"Failed to route to module: {target_module}",
                    "retryable": True
                },
                audit={
                    "module": "umbra-router",
                    "duration_ms": duration_ms
                }
            )
    
    @retry_async()
    async def _send_to_module(
        self,
        module_name: str,
        envelope: Envelope[UmbraPayload]
    ) -> Dict[str, Any]:
        """Send envelope to specific module."""
        module_config = self.modules[module_name]
        base_url = module_config["base_url"]
        api_key = module_config["api_key"]
        endpoint = f"{base_url}{module_config['api_endpoint']}"
        
        headers = {
            "Content-Type": "application/json",
            "X-Service-Name": "umbra"
        }
        
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"
        
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                endpoint,
                json=envelope.model_dump(),
                headers=headers
            )
            response.raise_for_status()
            return response.json()
    
    async def check_module_health(self, module_name: str) -> bool:
        """Check if a module is available and healthy."""
        if module_name not in self.modules:
            return False
        
        try:
            module_config = self.modules[module_name]
            health_url = f"{module_config['base_url']}{module_config['health_endpoint']}"
            
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(
                    health_url,
                    headers={"X-Service-Name": "umbra"}
                )
                return response.status_code == 200
                
        except Exception as e:
            self.logger.warning("Module health check failed",
                              module_name=module_name,
                              error=str(e))
            return False
    
    async def get_module_status(self) -> Dict[str, Dict[str, Any]]:
        """Get status of all available modules."""
        status = {}
        
        for module_name in self.modules.keys():
            start_time = datetime.utcnow()
            
            try:
                is_available = await self.check_module_health(module_name)
                latency = int((datetime.utcnow() - start_time).total_seconds() * 1000)
                
                status[module_name] = {
                    "available": is_available,
                    "latency": latency if is_available else None
                }
                
            except Exception as e:
                status[module_name] = {
                    "available": False,
                    "error": str(e)
                }
        
        return status
    
    async def route_with_fallback(
        self,
        envelope: Envelope[UmbraPayload],
        fallback_modules: Optional[list[str]] = None
    ) -> ModuleResult[Any]:
        """Route with fallback logic if primary module fails."""
        primary_module = envelope.payload.target_module
        
        if not primary_module:
            return await self.route_to_module(envelope)
        
        # Try primary module first
        result = await self.route_to_module(envelope)
        
        if result.status == "success":
            return result
        
        # Try fallback modules if provided
        if fallback_modules:
            self.logger.info("Trying fallback modules",
                           req_id=envelope.req_id,
                           primary_module=primary_module,
                           fallbacks=fallback_modules)
            
            for fallback_module in fallback_modules:
                if fallback_module == primary_module:
                    continue
                
                # Update envelope to target fallback module
                envelope.payload.target_module = fallback_module
                
                fallback_result = await self.route_to_module(envelope)
                if fallback_result.status == "success":
                    self.logger.info("Fallback routing successful",
                                   req_id=envelope.req_id,
                                   fallback_module=fallback_module)
                    return fallback_result
        
        # Return original error if all fallbacks failed
        return result